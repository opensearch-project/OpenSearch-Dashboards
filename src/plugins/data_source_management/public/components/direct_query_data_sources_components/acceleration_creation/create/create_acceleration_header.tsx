/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { EuiLink, EuiPageHeader, EuiPageHeaderSection, EuiSpacer, EuiText } from '@elastic/eui';
import React from 'react';
import { OPENSEARCH_ACC_DOCUMENTATION_URL } from '../../../constants';

export const CreateAccelerationHeader = () => {
  return (
    <div>
      <EuiPageHeader>
        <EuiPageHeaderSection>
          <EuiText size="s" data-test-subj="acceleration-header">
            <h1>Accelerate data</h1>
          </EuiText>
        </EuiPageHeaderSection>
      </EuiPageHeader>
      <EuiSpacer size="s" />
      <EuiText size="s" color="subdued">
        Create OpenSearch Indexes from external data connections for better performance.{' '}
        <EuiLink external={true} href={OPENSEARCH_ACC_DOCUMENTATION_URL} target="_blank">
          Learn more
        </EuiLink>
      </EuiText>
    </div>
  );
};
