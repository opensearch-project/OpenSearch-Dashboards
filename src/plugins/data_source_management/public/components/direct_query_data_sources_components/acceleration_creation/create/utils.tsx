/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import {
  ACCELERATION_INDEX_NAME_REGEX,
  ACCELERATION_S3_URL_REGEX,
} from '../../../../../framework/constants';
import {
  AccelerationIndexType,
  AccelerationRefreshType,
  CreateAccelerationForm,
  FormErrorsType,
  MaterializedViewQueryType,
  SkippingIndexRowType,
} from '../../../../../framework/types';

export const pluralizeTime = (timeWindow: number) => {
  return timeWindow > 1 ? 's' : '';
};

export const hasError = (formErrors: FormErrorsType, key?: keyof FormErrorsType) => {
  if (!key) return Object.values(formErrors).some((e) => !!e.length);
  return !!formErrors[key]?.length;
};

export const validateDataSource = (dataSource: string) => {
  return dataSource.trim().length === 0 ? ['Select a valid data source'] : [];
};

export const validateDatabase = (database: string) => {
  return database.trim().length === 0 ? ['Select a valid database'] : [];
};

export const validateDataTable = (dataTable: string) => {
  return dataTable.trim().length === 0 ? ['Select a valid table'] : [];
};

export const validatePrimaryShardCount = (primaryShardCount: number) => {
  return primaryShardCount < 1 ? ['Primary shards count should be greater than 0'] : [];
};

export const validateReplicaCount = (replicaCount: number) => {
  return replicaCount < 0 ? ['Replica count should be equal or greater than 0'] : [];
};

export const validateRefreshInterval = (
  refreshType: AccelerationRefreshType,
  refreshWindow: number,
  refreshInterval: string
) => {
  if (refreshType === 'autoInterval' && refreshInterval === 'minute' && refreshWindow < 15)
    return ['refresh window should be greater than 15 minutes'];

  if (refreshType === 'autoInterval' && refreshWindow < 1)
    return ['refresh window should be greater than 0'];

  return [];
};

export const validateWatermarkDelay = (
  accelerationIndexType: AccelerationIndexType,
  delayWindow: number
) => {
  return accelerationIndexType === 'materialized' && delayWindow < 1
    ? ['delay window should be greater than 0']
    : [];
};

export const validateIndexName = (value: string) => {
  // Check if the value contains lower case letters, numbers and underscore
  return !ACCELERATION_INDEX_NAME_REGEX.test(value) ? ['Enter a valid index name'] : [];
};

export const validateCheckpointLocation = (
  refreshType: AccelerationRefreshType,
  checkpointLocation: string | undefined
) => {
  if (refreshType !== 'manual' && !checkpointLocation) {
    return ['Checkpoint location is mandatory for auto refresh'];
  }

  if (checkpointLocation && !ACCELERATION_S3_URL_REGEX.test(checkpointLocation))
    return ['Enter a valid checkpoint location'];

  return [];
};

export const validateSkippingIndexData = (
  accelerationIndexType: AccelerationIndexType,
  skippingIndexQueryData: SkippingIndexRowType[]
) => {
  // TODO: Validate dataType match with supported acceleration method type
  if (accelerationIndexType !== 'skipping') return [];

  if (skippingIndexQueryData.length < 1) return ['Add fields to the skipping index definition'];

  return [];
};

export const validateCoveringIndexData = (
  accelerationIndexType: AccelerationIndexType,
  coveringIndexQueryData: string[]
) => {
  if (accelerationIndexType !== 'covering') return [];

  if (coveringIndexQueryData.length < 1) return ['Add fields to covering index definition'];
  return [];
};

export const validateMaterializedViewData = (
  accelerationIndexType: AccelerationIndexType,
  materializedViewQueryData: MaterializedViewQueryType
) => {
  if (accelerationIndexType !== 'materialized') return [];

  if (materializedViewQueryData.columnsValues.length < 1)
    return ['Add columns to materialized view definition'];

  if (materializedViewQueryData.groupByTumbleValue.timeField === '')
    return ['Add a time field to tumble function in materialized view definition'];

  if (materializedViewQueryData.groupByTumbleValue.tumbleWindow < 1)
    return ['Add a valid time window to tumble function in materialized view definition'];
  return [];
};

export const formValidator = (accelerationformData: CreateAccelerationForm) => {
  const {
    dataSource,
    database,
    dataTable,
    accelerationIndexType,
    skippingIndexQueryData,
    coveringIndexQueryData,
    materializedViewQueryData,
    accelerationIndexName,
    primaryShardsCount,
    replicaShardsCount,
    refreshType,
    checkpointLocation,
    watermarkDelay,
    refreshIntervalOptions,
  } = accelerationformData;

  const accelerationFormErrors: FormErrorsType = {
    dataSourceError: validateDataSource(dataSource),
    databaseError: validateDatabase(database),
    dataTableError: validateDataTable(dataTable),
    primaryShardsError: validatePrimaryShardCount(primaryShardsCount),
    replicaShardsError: validateReplicaCount(replicaShardsCount),
    refreshIntervalError: validateRefreshInterval(
      refreshType,
      refreshIntervalOptions.refreshWindow,
      refreshIntervalOptions.refreshInterval
    ),
    checkpointLocationError: validateCheckpointLocation(refreshType, checkpointLocation),
    watermarkDelayError: validateWatermarkDelay(accelerationIndexType, watermarkDelay.delayWindow),
    indexNameError: validateIndexName(accelerationIndexName),
    skippingIndexError: validateSkippingIndexData(accelerationIndexType, skippingIndexQueryData),
    coveringIndexError: validateCoveringIndexData(accelerationIndexType, coveringIndexQueryData),
    materializedViewError: validateMaterializedViewData(
      accelerationIndexType,
      materializedViewQueryData
    ),
  };

  return accelerationFormErrors;
};
