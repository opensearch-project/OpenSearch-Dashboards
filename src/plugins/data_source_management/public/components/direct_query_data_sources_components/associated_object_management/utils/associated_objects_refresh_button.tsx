/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { EuiButton } from '@elastic/eui';
import React from 'react';
import { ASSC_OBJ_REFRESH_BTN } from './associated_objects_tab_utils';

interface AssociatedObjectsRefreshButtonProps {
  isLoading: boolean;
  onClick: () => void;
}

export const AssociatedObjectsRefreshButton: React.FC<AssociatedObjectsRefreshButtonProps> = (
  props
) => {
  const { isLoading, onClick } = props;

  return (
    <EuiButton iconType="refresh" onClick={onClick} isLoading={isLoading} isDisabled={isLoading}>
      {ASSC_OBJ_REFRESH_BTN}
    </EuiButton>
  );
};
