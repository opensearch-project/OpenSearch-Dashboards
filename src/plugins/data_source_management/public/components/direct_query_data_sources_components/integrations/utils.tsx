/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { EuiHealth } from '@elastic/eui';
import React from 'react';

export const ASSET_FILTER_OPTIONS = [
  'index-pattern',
  'search',
  'visualization',
  'dashboard',
  'observability-search',
];

export const VALID_INDEX_NAME = /^[a-z\d\.][a-z\d\._\-\*]*$/;

export type Color = 'success' | 'primary' | 'warning' | 'danger' | undefined;

export const IntegrationHealthBadge = ({ status }: { status?: string }) => {
  switch (status) {
    case undefined:
      return <EuiHealth color="warning">Unknown</EuiHealth>;
    case 'available':
      return <EuiHealth color="success">Active</EuiHealth>;
    case 'partially-available':
      return <EuiHealth color="warning">Partially Available</EuiHealth>;
    default:
      return <EuiHealth color="danger">Critical</EuiHealth>;
  }
};
