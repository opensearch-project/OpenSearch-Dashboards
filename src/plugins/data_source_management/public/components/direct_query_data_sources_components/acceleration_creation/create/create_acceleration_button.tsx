/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { EuiButton } from '@elastic/eui';
import React, { useEffect, useState } from 'react';
import { HttpStart, NotificationsStart } from 'opensearch-dashboards/public';
import { SANITIZE_QUERY_REGEX } from '../../../../../framework/constants';
import {
  DirectQueryLoadingStatus,
  DirectQueryRequest,
  CreateAccelerationForm,
} from '../../../../../framework/types';
import { useDirectQuery } from '../../../../../framework/hooks/direct_query_hook';
import { accelerationQueryBuilder } from '../visual_editors/query_builder';
import { formValidator, hasError } from './utils';

interface CreateAccelerationButtonProps {
  accelerationFormData: CreateAccelerationForm;
  setAccelerationFormData: React.Dispatch<React.SetStateAction<CreateAccelerationForm>>;
  resetFlyout: () => void;
  refreshHandler?: () => void;
  http: HttpStart;
  notifications: NotificationsStart;
  dataSourceMDSId: string;
}

export const CreateAccelerationButton = ({
  accelerationFormData,
  setAccelerationFormData,
  resetFlyout,
  refreshHandler,
  http,
  notifications,
  dataSourceMDSId,
}: CreateAccelerationButtonProps) => {
  const { loadStatus: directqueryLoadStatus, startLoading: startDirectQuery } = useDirectQuery(
    http,
    notifications,
    dataSourceMDSId
  );
  const [isLoading, setIsLoading] = useState(false);

  const createAcceleration = () => {
    const errors = formValidator(accelerationFormData);
    if (hasError(errors)) {
      setAccelerationFormData({ ...accelerationFormData, formErrors: errors });
      return;
    }

    const requestPayload: DirectQueryRequest = {
      lang: 'sql',
      query: accelerationQueryBuilder(accelerationFormData).replaceAll(SANITIZE_QUERY_REGEX, ' '),
      datasource: accelerationFormData.dataSource,
    };

    startDirectQuery(requestPayload);
    setIsLoading(true);
  };

  useEffect(() => {
    const status = directqueryLoadStatus.toLowerCase();
    if (status === DirectQueryLoadingStatus.SUCCESS) {
      setIsLoading(false);
      notifications.toasts.addSuccess('Create acceleration query submitted successfully!');
      if (refreshHandler) refreshHandler();
      resetFlyout();
    } else if (
      status === DirectQueryLoadingStatus.FAILED ||
      status === DirectQueryLoadingStatus.CANCELED
    ) {
      setIsLoading(false);
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [directqueryLoadStatus]);

  const createAccelerationBtn = (
    <EuiButton onClick={createAcceleration} fill isLoading={isLoading}>
      {isLoading ? 'Creating acceleration' : 'Create acceleration'}
    </EuiButton>
  );

  return createAccelerationBtn;
};
