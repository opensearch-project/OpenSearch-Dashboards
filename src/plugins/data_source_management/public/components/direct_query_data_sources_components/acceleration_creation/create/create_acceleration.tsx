/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

/* eslint-disable react-hooks/exhaustive-deps */
/* eslint-disable no-console */

import {
  EuiButtonEmpty,
  EuiFlexGroup,
  EuiFlexItem,
  EuiFlyout,
  EuiFlyoutBody,
  EuiFlyoutFooter,
  EuiFlyoutHeader,
  EuiForm,
  EuiSpacer,
  htmlIdGenerator,
} from '@elastic/eui';
import React, { useEffect, useState } from 'react';
import { ApplicationStart, HttpStart, NotificationsStart } from 'opensearch-dashboards/public';
import {
  ACCELERATION_DEFUALT_SKIPPING_INDEX_NAME,
  ACCELERATION_REFRESH_TIME_INTERVAL,
  ACCELERATION_TIME_INTERVAL,
} from '../../../../../framework/constants';
import {
  CachedTable,
  CreateAccelerationForm,
  DirectQueryLoadingStatus,
} from '../../../../../framework/types';
import { useLoadTableColumnsToCache } from '../../../../../framework/catalog_cache/cache_loader';
import { CatalogCacheManager } from '../../../../../framework/catalog_cache/cache_manager';

import { IndexAdvancedSettings } from '../selectors/index_advanced_settings';
import { IndexSettingOptions } from '../selectors/index_setting_options';
import { IndexTypeSelector } from '../selectors/index_type_selector';
import { PreviewSQLDefinition } from '../selectors/preview_sql_defintion';
import { AccelerationDataSourceSelector } from '../selectors/source_selector';
import { QueryVisualEditor } from '../visual_editors/query_visual_editor';
import { CreateAccelerationButton } from './create_acceleration_button';
import { CreateAccelerationHeader } from './create_acceleration_header';
import { hasError } from './utils';

export interface CreateAccelerationProps {
  selectedDatasource: string;
  resetFlyout: () => void;
  databaseName?: string;
  tableName?: string;
  dataSourceMDSId?: string;
  refreshHandler?: () => void;
  http: HttpStart;
  notifications: NotificationsStart;
  application: ApplicationStart;
}

export const CreateAcceleration = ({
  selectedDatasource,
  resetFlyout,
  databaseName,
  tableName,
  dataSourceMDSId,
  refreshHandler,
  http,
  notifications,
  application,
}: CreateAccelerationProps) => {
  const [accelerationFormData, setAccelerationFormData] = useState<CreateAccelerationForm>({
    dataSource: selectedDatasource,
    database: databaseName ?? '',
    dataTable: tableName ?? '',
    dataTableFields: [],
    accelerationIndexType: 'skipping',
    skippingIndexQueryData: [],
    coveringIndexQueryData: [],
    materializedViewQueryData: {
      columnsValues: [],
      groupByTumbleValue: {
        timeField: '',
        tumbleWindow: 0,
        tumbleInterval: ACCELERATION_TIME_INTERVAL[2].value, // minutes
      },
    },
    accelerationIndexName: ACCELERATION_DEFUALT_SKIPPING_INDEX_NAME,
    primaryShardsCount: 1,
    replicaShardsCount: 1,
    refreshType: 'autoInterval',
    checkpointLocation: undefined,
    watermarkDelay: {
      delayWindow: 1,
      delayInterval: ACCELERATION_TIME_INTERVAL[2].value, // minutes
    },
    refreshIntervalOptions: {
      refreshWindow: 15,
      refreshInterval: ACCELERATION_REFRESH_TIME_INTERVAL[0].value, // minutes
    },
    formErrors: {
      dataSourceError: [],
      databaseError: [],
      dataTableError: [],
      skippingIndexError: [],
      coveringIndexError: [],
      materializedViewError: [],
      indexNameError: [],
      primaryShardsError: [],
      replicaShardsError: [],
      refreshIntervalError: [],
      checkpointLocationError: [],
      watermarkDelayError: [],
    },
  });
  const [tableFieldsLoading, setTableFieldsLoading] = useState(false);
  if (dataSourceMDSId === undefined) {
    dataSourceMDSId = '';
  }
  const {
    loadStatus,
    startLoading,
    stopLoading: stopLoadingTableFields,
  } = useLoadTableColumnsToCache(http, notifications);

  const loadColumnsToAccelerationForm = (cachedTable: CachedTable) => {
    const idPrefix = htmlIdGenerator()();
    const dataTableFields =
      cachedTable.columns?.map((col, index: number) => ({
        ...col,
        id: `${idPrefix}${index + 1}`,
      })) || [];

    setAccelerationFormData({
      ...accelerationFormData,
      dataTableFields,
    });
  };

  const initiateColumnLoad = (dataSource: string, database: string, dataTable: string) => {
    // All components related to table fields
    setAccelerationFormData({
      ...accelerationFormData,
      dataTableFields: [],
      skippingIndexQueryData: [],
      coveringIndexQueryData: [],
      materializedViewQueryData: {
        columnsValues: [],
        groupByTumbleValue: {
          timeField: '',
          tumbleWindow: 0,
          tumbleInterval: ACCELERATION_TIME_INTERVAL[2].value, // minutes
        },
      },
    });
    stopLoadingTableFields();
    if (dataTable !== '') {
      setTableFieldsLoading(true);
      try {
        const cachedTable = CatalogCacheManager.getTable(
          dataSource,
          database,
          dataTable,
          dataSourceMDSId
        );
        if (cachedTable.columns) {
          loadColumnsToAccelerationForm(cachedTable);
          setTableFieldsLoading(false);
        } else {
          startLoading({
            dataSourceName: dataSource,
            dataSourceMDSId,
            databaseName: database,
            tableName: dataTable,
          });
        }
      } catch (error) {
        notifications.toasts.addWarning('Your cache is outdated, refresh databases and tables');
        console.error(error);
      }
    }
  };

  useEffect(() => {
    if (databaseName !== undefined && tableName !== undefined) {
      initiateColumnLoad(
        accelerationFormData.dataSource,
        accelerationFormData.database,
        accelerationFormData.dataTable
      );
    }
  }, [databaseName, tableName]);

  useEffect(() => {
    const status = loadStatus.toLowerCase();
    if (status === DirectQueryLoadingStatus.SUCCESS) {
      let cachedTable = {} as CachedTable;
      try {
        cachedTable = CatalogCacheManager.getTable(
          accelerationFormData.dataSource,
          accelerationFormData.database,
          accelerationFormData.dataTable,
          dataSourceMDSId
        );
      } catch (error) {
        notifications.toasts.addWarning('Your cache is outdated, refresh databases and tables');
        console.error(error);
      }

      loadColumnsToAccelerationForm(cachedTable);
      setTableFieldsLoading(false);
    } else if (
      status === DirectQueryLoadingStatus.FAILED ||
      status === DirectQueryLoadingStatus.CANCELED
    ) {
      setTableFieldsLoading(false);
    }
  }, [loadStatus]);

  useEffect(() => {
    return () => {
      stopLoadingTableFields();
    };
  }, []);

  const dataSourcesPreselected = databaseName !== undefined && tableName !== undefined;

  return (
    <>
      <EuiFlyout ownFocus onClose={resetFlyout} aria-labelledby="flyoutTitle" size="m">
        <EuiFlyoutHeader hasBorder>
          <CreateAccelerationHeader />
        </EuiFlyoutHeader>
        <EuiFlyoutBody>
          <EuiForm
            isInvalid={hasError(accelerationFormData.formErrors)}
            error={Object.values(accelerationFormData.formErrors).flat()}
            component="div"
            id="acceleration-form"
          >
            <AccelerationDataSourceSelector
              http={http}
              notifications={notifications}
              accelerationFormData={accelerationFormData}
              setAccelerationFormData={setAccelerationFormData}
              selectedDatasource={selectedDatasource}
              dataSourcesPreselected={dataSourcesPreselected}
              tableFieldsLoading={tableFieldsLoading}
              dataSourceMDSId={dataSourceMDSId}
            />
            <EuiSpacer size="xxl" />
            <IndexTypeSelector
              accelerationFormData={accelerationFormData}
              setAccelerationFormData={setAccelerationFormData}
              initiateColumnLoad={initiateColumnLoad}
            />
            <EuiSpacer size="xxl" />
            <IndexSettingOptions
              accelerationFormData={accelerationFormData}
              setAccelerationFormData={setAccelerationFormData}
            />
            <EuiSpacer size="m" />
            <QueryVisualEditor
              accelerationFormData={accelerationFormData}
              setAccelerationFormData={setAccelerationFormData}
              tableFieldsLoading={tableFieldsLoading}
              dataSourceMDSId={dataSourceMDSId}
              http={http}
              notifications={notifications}
            />
            <EuiSpacer size="xxl" />
            <IndexAdvancedSettings
              accelerationFormData={accelerationFormData}
              setAccelerationFormData={setAccelerationFormData}
            />
            <EuiSpacer size="l" />
            <PreviewSQLDefinition
              accelerationFormData={accelerationFormData}
              setAccelerationFormData={setAccelerationFormData}
              resetFlyout={resetFlyout}
              notifications={notifications}
              application={application}
            />
          </EuiForm>
        </EuiFlyoutBody>
        <EuiFlyoutFooter>
          <EuiFlexGroup justifyContent="spaceBetween">
            <EuiFlexItem grow={false}>
              <EuiButtonEmpty iconType="cross" onClick={resetFlyout} flush="left">
                Cancel
              </EuiButtonEmpty>
            </EuiFlexItem>
            <EuiFlexItem grow={false}>
              <CreateAccelerationButton
                accelerationFormData={accelerationFormData}
                setAccelerationFormData={setAccelerationFormData}
                resetFlyout={resetFlyout}
                refreshHandler={refreshHandler}
                http={http}
                notifications={notifications}
                dataSourceMDSId={dataSourceMDSId}
              />
            </EuiFlexItem>
          </EuiFlexGroup>
        </EuiFlyoutFooter>
      </EuiFlyout>
    </>
  );
};
