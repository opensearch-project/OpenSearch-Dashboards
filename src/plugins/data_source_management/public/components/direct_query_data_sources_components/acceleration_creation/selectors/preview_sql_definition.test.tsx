/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from 'react';
import { mount } from 'enzyme';
import { act } from 'react-dom/test-utils';
import { PreviewSQLDefinition } from './preview_sql_definition';
import { EuiButton } from '@elastic/eui';
import { coreMock } from '../../../../../../../core/public/mocks';
import { formValidator, hasError } from '../create/utils';

jest.mock('../visual_editors/query_builder', () => ({
  accelerationQueryBuilder: jest.fn(() => 'mocked SQL query'),
}));

jest.mock('../create/utils', () => ({
  formValidator: jest.fn(() => ({})),
  hasError: jest.fn(() => false),
}));

describe('PreviewSQLDefinition', () => {
  const mockCoreStart = coreMock.createStart();
  const mockAccelerationFormData = {
    dataSource: 'test_source',
    formErrors: {},
  };

  const mockSetAccelerationFormData = jest.fn();
  const mockResetFlyout = jest.fn();

  beforeEach(() => {
    jest.clearAllMocks();
    mockCoreStart.http.get.mockResolvedValue({
      status: {
        statuses: [{ id: 'plugin:queryWorkbenchDashboards' }],
      },
    });
  });

  const mountComponent = async () => {
    const wrapper = mount(
      <PreviewSQLDefinition
        accelerationFormData={mockAccelerationFormData}
        setAccelerationFormData={mockSetAccelerationFormData}
        resetFlyout={mockResetFlyout}
        notifications={mockCoreStart.notifications}
        application={mockCoreStart.application}
        http={mockCoreStart.http}
      />
    );

    await act(async () => {
      await Promise.resolve();
    });
    wrapper.update();

    return wrapper;
  };

  test('renders component correctly', async () => {
    const wrapper = await mountComponent();
    expect(wrapper).toMatchSnapshot();
  });

  test('opens SQL Workbench when button is clicked', async () => {
    const wrapper = await mountComponent();

    const button = wrapper.findWhere(
      (node) => node.type() === EuiButton && node.prop('data-test-subj') === 'workbenchButton'
    );

    expect(button.exists()).toBe(true);
    expect(button.length).toBe(1);

    await act(async () => {
      button.prop('onClick')();
      await Promise.resolve();
    });

    expect(mockCoreStart.application.navigateToApp).toHaveBeenCalledWith(expect.any(String), {
      path: '#/test_source',
      state: {
        language: 'sql',
        queryToRun: 'mocked SQL query',
      },
    });
    expect(mockResetFlyout).toHaveBeenCalled();
  });

  test('displays error toast if SQL Workbench plugin check fails', async () => {
    mockCoreStart.http.get.mockRejectedValueOnce(new Error('Error checking plugin'));

    await mountComponent();

    expect(mockCoreStart.notifications.toasts.addDanger).toHaveBeenCalledWith(
      'Error checking Query Workbench Plugin Installation status.'
    );
  });

  test('shows update preview button when form data changes', async () => {
    const wrapper = await mountComponent();

    const previewButton = wrapper.findWhere(
      (node) => node.type() === EuiButton && node.text().includes('Generate preview')
    );

    await act(async () => {
      previewButton.prop('onClick')();
      await Promise.resolve();
    });

    await act(async () => {
      wrapper.setProps({
        accelerationFormData: { ...mockAccelerationFormData, dataSource: 'new_source' },
      });
      await Promise.resolve();
    });
    wrapper.update();

    const updateButton = wrapper.findWhere(
      (node) => node.type() === EuiButton && node.text().includes('Update preview')
    );

    expect(updateButton.exists()).toBe(true);
  });

  test('validates form before opening SQL Workbench', async () => {
    // Mock form validation to return errors
    (formValidator as jest.Mock).mockReturnValueOnce({ someField: 'error' });
    (hasError as jest.Mock).mockReturnValueOnce(true);

    const wrapper = await mountComponent();

    const button = wrapper.findWhere(
      (node) => node.type() === EuiButton && node.prop('data-test-subj') === 'workbenchButton'
    );

    await act(async () => {
      button.prop('onClick')();
      await Promise.resolve();
    });

    expect(mockCoreStart.application.navigateToApp).not.toHaveBeenCalled();
    expect(mockSetAccelerationFormData).toHaveBeenCalledWith({
      ...mockAccelerationFormData,
      formErrors: { someField: 'error' },
    });
  });

  test('hides SQL Workbench button when plugin is not installed', async () => {
    mockCoreStart.http.get.mockResolvedValueOnce({
      status: {
        statuses: [{ id: 'some-other-plugin' }],
      },
    });

    const wrapper = await mountComponent();

    const button = wrapper.findWhere(
      (node) => node.type() === EuiButton && node.prop('data-test-subj') === 'workbenchButton'
    );

    expect(button.exists()).toBe(false);
  });
});
