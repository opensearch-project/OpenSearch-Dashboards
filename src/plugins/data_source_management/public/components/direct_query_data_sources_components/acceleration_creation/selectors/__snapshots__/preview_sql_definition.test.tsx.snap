// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PreviewSQLDefinition renders component correctly 1`] = `
<PreviewSQLDefinition
  accelerationFormData={
    Object {
      "dataSource": "test_source",
      "formErrors": Object {},
    }
  }
  application={
    Object {
      "applications$": BehaviorSubject {
        "_isScalar": false,
        "_value": Map {},
        "closed": false,
        "hasError": false,
        "isStopped": false,
        "observers": Array [],
        "thrownError": null,
      },
      "capabilities": Object {
        "catalogue": Object {},
        "management": Object {},
        "navLinks": Object {},
        "workspaces": Object {},
      },
      "currentAppId$": Observable {
        "_isScalar": false,
        "source": Subject {
          "_isScalar": false,
          "closed": false,
          "hasError": false,
          "isStopped": false,
          "observers": Array [],
          "thrownError": null,
        },
      },
      "getUrlForApp": [MockFunction],
      "navigateToApp": [MockFunction],
      "navigateToUrl": [MockFunction],
      "registerMountContext": [MockFunction],
      "setAppBadgeControls": [MockFunction],
      "setAppBottomControls": [MockFunction],
      "setAppCenterControls": [MockFunction],
      "setAppDescriptionControls": [MockFunction],
      "setAppLeftControls": [MockFunction],
      "setAppRightControls": [MockFunction],
    }
  }
  http={
    Object {
      "addLoadingCountSource": [MockFunction],
      "anonymousPaths": Object {
        "isAnonymous": [MockFunction],
        "register": [MockFunction],
      },
      "basePath": BasePath {
        "basePath": "",
        "clientBasePath": "",
        "get": [Function],
        "getBasePath": [Function],
        "prepend": [Function],
        "remove": [Function],
        "serverBasePath": "",
      },
      "delete": [MockFunction],
      "fetch": [MockFunction],
      "get": [MockFunction] {
        "calls": Array [
          Array [
            "/api/status",
          ],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": Promise {},
          },
        ],
      },
      "getLoadingCount$": [MockFunction],
      "head": [MockFunction],
      "intercept": [MockFunction],
      "options": [MockFunction],
      "patch": [MockFunction],
      "post": [MockFunction],
      "put": [MockFunction],
    }
  }
  notifications={
    Object {
      "toasts": Object {
        "add": [MockFunction],
        "addDanger": [MockFunction],
        "addError": [MockFunction],
        "addInfo": [MockFunction],
        "addSuccess": [MockFunction],
        "addWarning": [MockFunction],
        "get$": [MockFunction],
        "remove": [MockFunction],
      },
    }
  }
  resetFlyout={[MockFunction]}
  setAccelerationFormData={[MockFunction]}
>
  <EuiAccordion
    arrowDisplay="left"
    buttonContent={
      <EuiText
        data-test-subj="preview-sql-header"
      >
        <h3>
          Preview SQL definition
        </h3>
      </EuiText>
    }
    id="accordion1"
    initialIsOpen={false}
    isLoading={false}
    isLoadingMessage={false}
    paddingSize="l"
  >
    <div
      className="euiAccordion"
    >
      <div
        className="euiAccordion__triggerWrapper"
      >
        <button
          aria-controls="accordion1"
          aria-expanded={false}
          className="euiAccordion__button"
          id="generated-id"
          onClick={[Function]}
          type="button"
        >
          <span
            className="euiAccordion__iconWrapper"
          >
            <EuiIcon
              className="euiAccordion__icon"
              size="m"
              type="arrowRight"
            >
              <span
                className="euiAccordion__icon"
                data-euiicon-type="arrowRight"
                size="m"
              />
            </EuiIcon>
          </span>
          <span
            className="euiIEFlexWrapFix"
          >
            <EuiText
              data-test-subj="preview-sql-header"
            >
              <div
                className="euiText euiText--medium"
                data-test-subj="preview-sql-header"
              >
                <h3>
                  Preview SQL definition
                </h3>
              </div>
            </EuiText>
          </span>
        </button>
      </div>
      <div
        aria-labelledby="generated-id"
        className="euiAccordion__childWrapper"
        id="accordion1"
        role="region"
        tabIndex={-1}
      >
        <EuiResizeObserver
          onResize={[Function]}
        >
          <div>
            <div
              className="euiAccordion__padding--l"
            >
              <EuiFlexGroup>
                <div
                  className="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--directionRow euiFlexGroup--responsive"
                >
                  <EuiFlexItem
                    grow={false}
                  >
                    <div
                      className="euiFlexItem euiFlexItem--flexGrowZero"
                    >
                      <EuiButton
                        color="success"
                        onClick={[Function]}
                      >
                        <EuiButtonDisplay
                          baseClassName="euiButton"
                          color="success"
                          disabled={false}
                          element="button"
                          isDisabled={false}
                          onClick={[Function]}
                          type="button"
                        >
                          <button
                            className="euiButton euiButton--success"
                            disabled={false}
                            onClick={[Function]}
                            style={
                              Object {
                                "minWidth": undefined,
                              }
                            }
                            type="button"
                          >
                            <EuiButtonContent
                              className="euiButton__content"
                              iconGap="m"
                              iconSide="left"
                              textProps={
                                Object {
                                  "className": "euiButton__text",
                                }
                              }
                            >
                              <span
                                className="euiButtonContent euiButton__content"
                              >
                                <span
                                  className="euiButton__text"
                                >
                                  Generate preview
                                </span>
                              </span>
                            </EuiButtonContent>
                          </button>
                        </EuiButtonDisplay>
                      </EuiButton>
                    </div>
                  </EuiFlexItem>
                  <EuiFlexItem
                    grow={false}
                  >
                    <div
                      className="euiFlexItem euiFlexItem--flexGrowZero"
                    >
                      <EuiButton
                        data-test-subj="workbenchButton"
                        iconSide="right"
                        onClick={[Function]}
                      >
                        <EuiButtonDisplay
                          baseClassName="euiButton"
                          data-test-subj="workbenchButton"
                          disabled={false}
                          element="button"
                          iconSide="right"
                          isDisabled={false}
                          onClick={[Function]}
                          type="button"
                        >
                          <button
                            className="euiButton euiButton--primary"
                            data-test-subj="workbenchButton"
                            disabled={false}
                            onClick={[Function]}
                            style={
                              Object {
                                "minWidth": undefined,
                              }
                            }
                            type="button"
                          >
                            <EuiButtonContent
                              className="euiButton__content"
                              iconGap="m"
                              iconSide="right"
                              textProps={
                                Object {
                                  "className": "euiButton__text",
                                }
                              }
                            >
                              <span
                                className="euiButtonContent euiButtonContent--iconRight euiButton__content"
                              >
                                <span
                                  className="euiButton__text"
                                >
                                  Edit in Query Workbench
                                </span>
                              </span>
                            </EuiButtonContent>
                          </button>
                        </EuiButtonDisplay>
                      </EuiButton>
                    </div>
                  </EuiFlexItem>
                </div>
              </EuiFlexGroup>
              <EuiSpacer
                size="l"
              >
                <div
                  className="euiSpacer euiSpacer--l"
                />
              </EuiSpacer>
              <EuiCodeBlock
                fontSize="m"
                isCopyable={true}
                language="sql"
                paddingSize="m"
              >
                <EuiCodeBlockImpl
                  fontSize="m"
                  inline={false}
                  isCopyable={true}
                  language="sql"
                  paddingSize="m"
                >
                  <div>
                    <pre>
                      <code />
                    </pre>
                  </div>
                </EuiCodeBlockImpl>
              </EuiCodeBlock>
            </div>
          </div>
        </EuiResizeObserver>
      </div>
    </div>
  </EuiAccordion>
</PreviewSQLDefinition>
`;
